firedrake/mesh.py

```diff
@@ -1537,6 +1686,175 @@ def ExtrudedMesh(mesh, layers, layer_height=None, extrusion_type='uniform', kern
+    element = ufl.VectorElement("DG", cell, 0)
+    # Create mesh object
+    vmesh = MeshGeometry.__new__(MeshGeometry, element)
+    vmesh._topology = topology
+    vmesh._parent_mesh = mesh
+
+    def callback(self):
+        """Finish initialisation."""
+        del self._callback
+        # Finish the initialisation of mesh topology
+        self.topology.init()
+
+        coordinates_fs = functionspace.VectorFunctionSpace(self.topology, "DG", 0,
+                                                           dim=gdim)
+
+        coordinates_data = dmswarm.reordered_coords(swarm, coordinates_fs.dm.getDefaultSection(),
+                                                    (self.num_vertices(), gdim))
+
+        coordinates = function.CoordinatelessFunction(coordinates_fs,
+                                                      val=coordinates_data,
+                                                      name="Coordinates")
+
+        self.__init__(coordinates)
+
+    vmesh._callback = callback
+    return vmesh
+
+
+def _pic_swarm_in_plex(plex, coords, **kwargs):
+    """
+    Create a Particle In Cell (PIC) DMSwarm, immersed in a DMPlex
+    at given point coordinates.
+
+    This should only by used for dmplexes associated with meshes with
+    straight edges. If not, the particles may be placed in the wrong
+    cells.
+
+    :arg plex: the DMPlex within with the DMSwarm should be
+        immersed.
+    :arg coords: an `ndarray` of (npoints, coordsdim) shape.
+    :kwarg fields: An optional list of named data which can be stored
+        for each point in the DMSwarm. The format should be
+        `[(fieldname1, blocksize1, dtype1),
+          ...,
+         (fieldnameN, blocksizeN, dtypeN)]`.
+        For example, the swarm coordinates themselves are stored in a
+        field named `DMSwarmPIC_coor` which, were in not created
+        automatically, would be initialised with
+        `fields = [("DMSwarmPIC_coor", coordsdim, ScalarType)]`.
+        All fields must have the same number of points. For more
+        information see https://www.mcs.anl.gov/petsc/petsc-current/docs/manualpages/DMSWARM/DMSWARM.html
+    :return: the immersed DMSwarm
+
+    .. note::
+
+        The created DMSwarm uses the communicator of the input DMPlex.
+    """
+
+    coords = np.asarray(coords, dtype=np.double)
+
+    # Create a DMSWARM
+    swarm = PETSc.DMSwarm().create(comm=plex.comm)
+
+    # Set swarm DM dimension to match DMPlex dimension
+    # NB: Unlike a DMPlex, this does not correspond to the topological
+    #     dimension of a mesh (which would be 0). In all PETSc examples
+    #     the dimension of the DMSwarm is set to match that of the
+    #     DMPlex used with swarm.setCellDM
+    swarm.setDimension(plex.getDimension())
+
+    # Set coordinates dimension
+    npoints, coordsdim = coords.shape
+    swarm.setCoordinateDim(coordsdim)
+
+    if coordsdim == 1:
+        raise NotImplementedError("1D DMSwarm not yet supported")
```

Comment:
Move this error before making things. Also, is it not supported by us, or not supported by them?

Reply:
PETSc. You can't use a 1D Plex as cell DM. I'll update the message

Reply:
Is it DMSwarm, or DMLocateCell?

Reply:
With `test.py` containing
```python
from firedrake import *
m = UnitIntervalMesh(1)
vertices = [(0.,), (0.1,), (0.9,), (1.0,)]
vm = VertexOnlyMesh(m, vertices)
```
I get
```
Traceback (most recent call last):
  File "test.py", line 4, in <module>
    vm = VertexOnlyMesh(m, vertices)
  File "/Users/rwh10/firedrake/src/firedrake/firedrake/mesh.py", line 1733, in VertexOnlyMesh
    swarm = _pic_swarm_in_plex(mesh.topology._plex, vertexcoords, fields=[("parentcellnum", 1, IntType)])
  File "/Users/rwh10/firedrake/src/firedrake/firedrake/mesh.py", line 1846, in _pic_swarm_in_plex
    swarm.setPointCoordinates(coords, redundant=False, mode=PETSc.InsertMode.INSERT_VALUES)
  File "PETSc/DMSwarm.pyx", line 200, in petsc4py.PETSc.DMSwarm.setPointCoordinates
petsc4py.PETSc.Error: error code 63
[0] DMSwarmSetPointCoordinates() line 305 in /Users/rwh10/firedrake/src/petsc/src/dm/impls/swarm/swarmpic.c
[0] DMLocatePoints() line 6550 in /Users/rwh10/firedrake/src/petsc/src/dm/interface/dm.c
[0] DMLocatePoints_Plex() line 738 in /Users/rwh10/firedrake/src/petsc/src/dm/impls/plex/plexgeometry.c
[0] DMPlexLocatePoint_Internal() line 460 in /Users/rwh10/firedrake/src/petsc/src/dm/impls/plex/plexgeometry.c
[0] Argument out of range
[0] No point location for cell 0 with type segment
```

Reply:
OK, so if you wanted to support this you could implement `DMPlexLocatePoint_Simplex_1D_Internal` and hook it up in `DMPlexLocatePoint_Internal`

Reply:
See issue #1792